### ISSUE TITLE
ServerVersion.AutoDetect hangs when connecting to RDS Proxy

### ISSUE THREAD

**Issue by mertkokusen** [PomeloFoundation/Pomelo.EntityFrameworkCore.MySql#1736]
### Steps to reproduce
Trying to connect to RDS Proxy for MySql while using `ServerVersion.AutoDetect(rdsProxyConnectionString)`.

### The issue
When using AutoDetect for server version, app hangs indefinetly when it tries to connect to the database proxy. No exceptions are thrown and no errors are logged. 
If you replace `ServerVersion.AutoDetect(rdsProxyConnectionString)` for example with `ServerVersion.Parse("8.0.28")` app connects to the proxy without any problems.

### Further technical details

MySQL version: 8.0.28
Operating system: Windows / Linux
Pomelo.EntityFrameworkCore.MySql version: 6.0.2
Microsoft.AspNetCore.App version: 6.0.4

**Comment by mguinness** [comment]
Assuming that you're using Amazon RDS for MySQL as the backend database server and not Amazon Aurora.

The AutoDetect code is very simple, it connects to the instance using [MySqlConnector](https://mysqlconnector.net/) and parses the value from the [ServerVersion](https://mysqlconnector.net/api/mysqlconnector/mysqlconnection/serverversion/) property.

https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/blob/00b4353d95a342bed33f8a15b63f9afb02933a39/src/EFCore.MySql/Infrastructure/ServerVersion.cs#L71-L77

Try replicating that in a test console app using [MySqlConnector 2.1.2](https://www.nuget.org/packages/MySqlConnector/2.1.2) (used by Pomelo 6.0.2) and configure [Logging](https://mysqlconnector.net/overview/logging/) set to trace log level.

**Comment by jasonterando** [comment]
I ran into something similar with connecting to Aurora directly.  Try adding `;Pipelining=False` to your connection string and see if that helps.

**Comment by aaron-ballard-530** [comment]
I'm running into this issue as well on an MySQL Aurora 5.7 cluster.  When I have the **Server** value of the connection string pointed to an AWS Route53 private zone the first connection works but the next request just hangs as described above.  Pointing to the public writer and reader CNAMEs provided by AWS for the cluster are working without issue.  

The implementation is using the AutoDetect setting and I've also tried setting it explicitly to 5.7.  

Runtime: .net 6
Pomelo: 6.0.1
MySqlConnector 2.1.2
EFCore.Relational 6.0.1

Setting the pipelining to false seems to correct the issue however it doesn't make any sense that it the failure is tied to just a private CNAME record.  According to this https://mysqlconnector.net/troubleshooting/aurora-freeze/ it should always fail.

Has a correction been done in version 7.x to handle the issue?

**Comment by jeyamaal** [comment]
Hi team,

Has the issue been fixed, or is someone currently working on it ?

I am also encountering the same issue.

My existing setup with **RDS Proxy without IAM Authentication** works with `ServerVersion.AutoDetect(rdsProxyConnectionString)`, but it does not work with IAM Authentication.

It functions correctly when I use @mertkokusen's solution `ServerVersion.Parse("8.0.28")` for RDS Proxy with IAM Authentication.

However, I need to use `ServerVersion.AutoDetect` to pass the connection string, as it would be cumbersome to manually update it every time a new update is released.

Runtime: .net 8
Pomelo: 8.0.2
EFCore.Relational 8.0.7
EntityFrameworkCore 8.0.7
EFCore.Tools 8.0.7
Engine version : 8.0.mysql_aurora.3.07.1

**Comment by ljsomm** [comment]
Up

**Comment by lauxjpn** [comment]
@jeyamaal @ljsomm Did you try @jasonterando proposed solution?

> I ran into something similar with connecting to Aurora directly. Try adding `;Pipelining=False` to your connection string and see if that helps.

It worked for @nccorro in https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/issues/1749#issuecomment-1405053787 as well.

---

If that doesn't work, take a look at the implementation of `ServerVersion.AutoDetect()` (which is really quite simple), copy it into a new method and modify it until it works for you:

https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/blob/b1efe54a18a8eb61a2c4dd09c121e6191e0730e5/src/EFCore.MySql/Infrastructure/ServerVersion.cs#L73-L84

Since it really just opens a simple connection, you would need to troubleshoot connecting to the database server in general, if that doesn't work.

**Comment by ljsomm** [comment]
@lauxjpn 

Yeah, I tried.
It finely works directly for Aurora but whenever I try to do this with the proxy it doesn't work.
The problem is not related to any infrastructure components (such as security groups) because I already did a troubleshoot from the app to the proxy using `mysql cli` and it finely works.

The only way that actually works was by defining the specific version, but that's not what we should look for considering that every time we have the database engine updated, we will also need to increase the version in the code.


A simple summary of everything said here:

- ServerVersion.AutoDetect() works only with the Aurora Instance, but not the proxy, even if we are using `;Pipelining=False`.

- Specify the version works in both scenarios, but I believe that's not what @jeyamaal and I've been looking for.

**Comment by lauxjpn** [comment]
@ljsomm As you can see by the `ServerVersion.AutoDetect()` code I posted, it sets/modifies 3 options of your connection string and then just opens a connection with it.

So if your App works when not using `ServerVersion.AutoDetect()`, then either at least one of the options set by `ServerVersion.AutoDetect()` is the issue, or you somehow exhausted all available connections of either the database server or the proxy.

I would start by copying the code, I posted above, to your app, then change your app to call that copy instead of the code that ships with Pomelo, and then remove the connection string options one by one, testing the app afterwards every time.

**Comment by ljsomm** [comment]
@lauxjpn that's correct, the problem is related to the Empty string that is given by Pomelo implementation.
